src/session/content/background.rs

```diff
@@ -0,0 +1,397 @@
+        } else {
+            float alpha = pattern.a * 0.1;
+            pattern_color = vec4(vec3(0.0), 1.0) * alpha;
+        }
+
+        // blend colors with premultiplied alpha
+        fragColor = messages + pattern_color * (1.0 - messages.a);
+    }
+}
+"#
+.as_bytes();
+
+mod imp {
+    use super::*;
+    use gtk::glib::once_cell::sync::{Lazy, OnceCell};
+    use gtk::glib::subclass::Signal;
+    use gtk::glib::{self, clone};
+
+    #[derive(PartialEq, Clone, Copy)]
+    struct GradientCacheParams {
+        size: (f32, f32),
+        colors: [graphene::Vec3; 4],
+        phase: u32,
+    }
+
+    #[derive(Default)]
+    pub struct Background {
+        pub(super) gradient_cache: RefCell<Option<gdk::Texture>>,
+        pub(super) last_size: Cell<(f32, f32)>,
+
+        pub(super) shaders: RefCell<Option<Option<[gsk::GLShader; 2]>>>,
+        pub(super) pattern: OnceCell<gdk::Texture>,
+
+        pub(super) progress: Cell<f32>,
+        pub(super) phase: Cell<u32>,
+
+        pub(super) dark: Cell<bool>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Background {
+        const NAME: &'static str = "ComponentsBackground";
+        type Type = super::Background;
+        type ParentType = adw::Bin;
+    }
+
+    impl ObjectImpl for Background {
+        fn constructed(&self) {
+            self.parent_constructed();
+
+            let obj = self.obj();
+
+            obj.set_vexpand(true);
+            obj.set_hexpand(true);
+
+            let pattern =
+                gdk::Texture::from_resource("/com/github/melix99/telegrand/images/pattern.svg");
+
+            self.pattern.set(pattern).unwrap();
+
+            let style_manager = adw::StyleManager::default();
+            self.dark.set(style_manager.is_dark());
+
+            style_manager.connect_dark_notify(clone!(@weak obj => move |style_manager| {
+                let imp = obj.imp();
+                imp.dark.set(style_manager.is_dark());
+                imp.gradient_cache.take();
+            }));
+
+            let target = adw::CallbackAnimationTarget::new(clone!(@weak obj => move |progress| {
+                let imp = obj.imp();
+                imp.gradient_cache.take();
+                let progress = progress as f32;
+                if progress >= 1.0 {
+                    imp.progress.set(0.0);
+                    imp.phase.set((imp.phase.get() + 1) % 8);
+                } else {
+                    imp.progress.set(progress)
+                }
+                obj.queue_draw();
+            }));
+
+            let animation = adw::TimedAnimation::new(obj.as_ref(), 0.0, 1.0, 200, &target);
+            animation.set_easing(adw::Easing::EaseInOutQuad);
+
+            obj.connect_local("animate", true, move |_| {
+                let val = animation.value();
+                if val == 0.0 || val == 1.0 {
+                    animation.play()
+                }
+                None
+            });
```

Comment:
This is weird. Isn't it better to just have an "animate" method and call this inside it?

Reply:
The signal doesn't seem something we really need.